### ISSUE TITLE
libuv tty reopening interferes with native fork/exec

### ISSUE THREAD

**Issue by anszom** [libuv/libuv#2062]
* **Version**: 1.20.3 
* **Platform**: Linux (also applies to other unix-like OS)

After libuv tty code touches the stdin/stdout/stderr streams, it reopens them with the FD_CLOEXEC flag. Any further attempts to spawn child processes using the native unix fork/exec mechanism will result with the children having the standard FD's closed. The application will need to use uv_spawn to have these FD's retained in child processes.

Minimal repro case:
```
#include <uv.h>
#include <stdlib.h>

int main(int ac, char ** av)
{
        uv_loop_t loop;
        uv_tty_t fd1;
        uv_loop_init(&loop);
        uv_tty_init(&loop, &fd1, 1, 0);
        system("echo this should appear on stdout");
        return 0;
}
```

This causes a real-world bug in [Electron]( https://github.com/electron/electron) (https://github.com/electron/electron/issues/14246, https://github.com/electron/electron/issues/12937) where the chromium code is linked together with libuv in node.js. Chromium is unaware of libuv and attempts to natively spawn a gpu worker subprocess, which starts with fds 0-2 closed. These free fds are assigned for other purposes, and on some platforms the OpenGL blindly prints some warning messages to fd 2, messing up the state of the application.

**Comment by bnoordhuis** [comment]
> The application will need to use uv_spawn to have these FD's retained in child processes.

That's correct and intentional: libuv reopens the file descriptor when it refers to a slave pty, otherwise other processes that share the file descriptor (think `node | cat`) are affected when libuv switches it to non-blocking mode.

It probably needs to be documented better but the current behavior is not a bug. Pull request welcome, of course. :-)

**Comment by anszom** [comment]
> That's correct and intentional: libuv reopens the file descriptor when it refers to a slave pty, otherwise other processes that share the file descriptor (think node | cat) are affected when libuv switches it to non-blocking mode.

That is clear to me. I'm not sure about the intentional part. Do you mean that using libuv's tty functions is allowed only when an app doesn't perform native fork/exec?

> Pull request welcome, of course. :-)

For this I would need a better understanding of the expected semantics. My first idea would be to keep the tty-reopening code, but simply keep the old fd untouched and use the *new, reopened* fd in libuv. At least for fd's 0,1,2. Would this break anything?

**Comment by bnoordhuis** [comment]
> Do you mean that using libuv's tty functions is allowed only when an app doesn't perform native fork/exec?

Pretty much. The interaction between stdio and process management on Unices is such that you can't reliably do one without also controlling the other.

Sorry if I was unclear: "pull request" means "documentation pull request."

> simply keep the old fd untouched and use the new, reopened fd in libuv. At least for fd's 0,1,2. Would this break anything?

In downstream projects, you mean? Almost certainly.

**Comment by vtjnash** [comment]
We've been tracking a variant of this issue for a while time on the Julia project (https://github.com/JuliaLang/julia/issues/24440), where libuv also violates the internal assumptions about its re-opened file descriptor (those described above by bnoordhuis) if you do choose to go through the uv_spawn API.

@anszom as a workaround, you can always pass `dup(1)` instead (This is what JuliaLang does. That also has the benefit that native C stdio also keeps working in the current process (which normally doesn't handle O_NONBLOCK very well)—e.g. the same pipeline issue mentioned above, but within the same process.)

**Comment by anszom** [comment]
Very well, if you insist that this behaviour is not a bug, then it clearly needs to be documented. Silently marking stdio streams as FD_CLOEXEC is not very friendly.

@bnoordhuis, can you suggest any workarounds? vtjnash suggests passing dup(1) to tty code, which should - as I understand it - have the same effect as the modification that I've described above. What are the downsides? You mentioned that this would likely break some downstream projects, I'm not sure what you had in mind.

**Comment by bnoordhuis** [comment]
@anszom There are programs out there that use libuv for tty management but poll the stdio fds themselves.

The change you proposed would break that, with no easy fix for such programs except mirroring libuv's tty reopen logic. That wouldn't help e.g. a node.js script that calls `fs.read(0, cb)`, since it has no direct control over that, it's managed by the node.js run-time.

If electron doesn't fall in that category, then dup is probably an effective solution.

**Comment by anszom** [comment]
Electron simply embeds node.js, so a sufficiently motivated user could trigger this issue.

For now I've suggested a workaround which simply reopens /dev/null as stdio at startup (if missing), but it's an ugly hack to be honest.

**Comment by vtjnash** [comment]
> There are programs out there that use libuv for tty management but poll the stdio fds themselves.
> The change you proposed would break that, with no easy fix for such programs except mirroring libuv's tty reopen logic.

I don't think these statements necessarily follow. In Julia, we `dup` the fd specifically to prevent this (among other reasons), because avoiding modification of the original stdio handle can actually _improve_ the ability for programs to use the original stdio fds outside of libuv. Since in the kernel they still represent the same file, `poll` and `fs.read` should still continue to work, but additionally naive IO (such as `std::cout` or `puts`, or external `fork/exec`, and `uv_spawn`) also continue to work within the same program.

> That wouldn't help e.g. a node.js script that calls fs.read(0, cb), since it has no direct control over that, it's managed by the node.js run-time.

It sounds like you're asserting that libuv does not have direct control over fd 0 and that it should be managed by the node.js run-time. But here we see that libuv _is_ asserting control over fd 0, so that simply attempting to load up node.js as a library is enough to break the parent application. I think that is precisely the problem anszom is describing!

But, to be sure, I don't think this is actually a libuv bug, and so it may be preferable to address this in node.js. It would be a significantly breaking change for libuv to stop taking ownership of the `fd` it was handed (since this would need to be done uniformly across all APIs).

**Comment by bnoordhuis** [comment]
I'm talking about `uv_tty_set_mode()`. I wouldn't want to vouch that using a dup'd file descriptor always works. We've had regressions in the past in this area (in Node.js) caused by libuv changes.

> It sounds like you're asserting that libuv does not have direct control

Not sure I follow. The "it" in "it has no direct control" refers to the JS script, not libuv. What happens under the hood is mostly a black box to a script.

**Comment by vtjnash** [comment]
Ah, interesting. Do you have a reference for which issues those would be? I'm curious and would like to understand this better. I would assume a `dup` file descriptor should work exactly the same as the original, by virtue of having got the handle originally via `fork` making a copy of it when the shell started us. I would not be too surprised to learn that some aspect of the mode changes weren't properly handled for the re-opened file vs. the original, even though those changes should be properly reflected (since, e.g., [stty](https://linux.die.net/man/1/stty) can see/change them completely externally).

Note that there's already precedence in libuv to treating fd 0-2 specially: if you attempt to close this fd number, libuv won't actually close the underlying fd. That could argue in favor of this proposal to not add the CLOEXEC flag for 0-2 as well. (Although personally, I still recommend that applications simply avoid passing those handles to libuv, since libuv is expected to take ownership of the fd, but it should not—and generally tries to avoid—taking ownership of the special stdio handles.)

> Not sure I follow

Oh, I think we're talking past each other. I was considering the effect on an application that was using node.js as an embedded language.

> That wouldn't help e.g. a node.js script that calls fs.read(0, cb), since it has no direct control over that, it's managed by the node.js run-time.

Yikes, this sounds like it may actually be a significant bug in the runtime which may lead to data loss. The `fs.c` code currently mandates that file handles must be opened in blocking mode, since the fix for this data loss (e.g. PR #640 that we just got merged) depends on `write` blocking until it has written at least one byte (or that all future writes should fail). I'm not actually sure what would happen in the `fs.read` case, since I think that would depend on how the user handled it further up the stack. Presumably they've got a `select` call there, so doing a non-blocking read, to get all available data, is the intended result. For the writes to be correct, it would appear that we need to add a call to `select` there (in the threadpool). If you agree, I can fork this comment into a new issue here using GitHub's new "create issue button" :)

**Comment by vtjnash** [comment]
Hm, I'm not sure this is stale, although it also might be hard to untangle this. It's possible I should just write documentation on how we workaround this in JuliaLang mentioned above? (not sure that'll help OP though, since, if I understand correctly, that's a transitive issue from Electron using nodejs and they don't have the direct ability to prevent nodejs from modifying fd 0-2)
